lib/FrameManager.js

```diff
@@ -581,22 +588,64 @@ class Frame {
    * @return {!Promise}
    */
   waitForSelector(selector, options = {}) {
+    return this._waitForSelectorOrXpath(selector, false, options);
+  }
+
+  /**
+   * @param {string} xpath
+   * @param {!Object=} options
+   * @return {!Promise}
+   */
+  waitForXpath(xpath, options = {}) {
+    return this._waitForSelectorOrXpath(xpath, true, options);
+  }
+
+  /**
+   * @param {Function|string} pageFunction
+   * @param {!Object=} options
+   * @return {!Promise}
+   */
+  waitForFunction(pageFunction, options = {}, ...args) {
+    const timeout = options.timeout || 30000;
+    const polling = options.polling || 'raf';
+    return new WaitTask(this, pageFunction, polling, timeout, ...args).promise;
+  }
+
+  /**
+   * @return {!Promise<string>}
+   */
+  async title() {
+    return this.evaluate(() =>  document.title);
+  }
+
+  /**
+   * @param {string} selectorOrXpath
+   * @param {boolean} isXpath
+   * @param {!Object=} options
+   * @return {!Promise}
+   */
+  _waitForSelectorOrXpath(selectorOrXpath, isXpath, options = {}) {
     const timeout = options.timeout || 30000;
     const waitForVisible = !!options.visible;
     const waitForHidden = !!options.hidden;
     const polling = waitForVisible || waitForHidden ? 'raf' : 'mutation';
-    return this.waitForFunction(predicate, {timeout, polling}, selector, waitForVisible, waitForHidden);
+    return this.waitForFunction(predicate, {timeout, polling}, selectorOrXpath, isXpath, waitForVisible, waitForHidden);
 
     /**
-     * @param {string} selector
+     * @param {string} selectorOrXpath
+     * @param {boolean} isXpath
      * @param {boolean} waitForVisible
      * @param {boolean} waitForHidden
      * @return {?Node|boolean}
      */
-    function predicate(selector, waitForVisible, waitForHidden) {
-      const node = document.querySelector(selector);
+    function predicate(selectorOrXpath, isXpath, waitForVisible, waitForHidden) {
+      const node = /** @type {Element} */ (isXpath
+        ? document.evaluate(selectorOrXpath, document, null, XPathResult.FIRST_ORDERED_NODE_TYPE, null).singleNodeValue
+        : document.querySelector(selectorOrXpath));
       if (!node)
         return waitForHidden;
+      if (isXpath && node.nodeType !== Node.ELEMENT_NODE)
+        throw 'waitForXpath only supports element selection';
```

Comment:
I think these days ElementHandle is actually fine to use for nodes. We use it internally as a reference to the Document node. So this check can be removed.

Reply:
Alright, fixed this code to work with text nodeds and adding a test to verify selecting textNode works